noir-projects/aztec-nr/aztec/src/messages/encryption/aes128.nr

```diff
@@ -82,35 +84,47 @@ fn extract_close_to_uniformly_random_256_bits_from_ecdh_shared_secret_using_pose
     // not uniformly random over the byte: it only has < 6 bits of randomness, because
     // it's the big end of a 254-bit field element.
 
-    let rand1: Field = poseidon2_hash_with_separator(
-        [shared_secret.x, shared_secret.y],
-        GENERATOR_INDEX__SYMMETRIC_KEY,
-    );
-    let rand2: Field = poseidon2_hash_with_separator(
-        [shared_secret.x, shared_secret.y],
-        GENERATOR_INDEX__SYMMETRIC_KEY_2,
-    );
-
-    let rand1_bytes: [u8; 32] = rand1.to_be_bytes();
-    let rand2_bytes: [u8; 32] = rand2.to_be_bytes();
-
-    let mut bytes: [u8; 32] = [0; 32];
-    for i in 0..16 {
-        // We take bytes from the "little end" of the be-bytes arrays:
-        let j = 32 - i - 1;
-        bytes[i] = rand1_bytes[j];
-        bytes[16 + i] = rand2_bytes[j];
+    let mut all_bytes: [[u8; 32]; N] = std::mem::zeroed();
+
+    for k in 0..N {
+        assert(N < 256); // TODO: does this incur a range constraint, or is it comptime?
+
+        // We augment the domain separator with the loop index, so that we can
+        // generate N lots of randomness.
+        let k_shift = (k as u16 << 8);
+        let separator_1 = k_shift + GENERATOR_INDEX__SYMMETRIC_KEY as u16;
+        let separator_2 = k_shift + GENERATOR_INDEX__SYMMETRIC_KEY_2 as u16;
+
+        let rand1: Field =
+            poseidon2_hash_with_separator([shared_secret.x, shared_secret.y], separator_1);
+        let rand2: Field =
+            poseidon2_hash_with_separator([shared_secret.x, shared_secret.y], separator_2);
+
+        let rand1_bytes: [u8; 32] = rand1.to_be_bytes();
+        let rand2_bytes: [u8; 32] = rand2.to_be_bytes();
+
+        let mut bytes: [u8; 32] = [0; 32];
+        for i in 0..16 {
+            // We take bytes from the "little end" of the be-bytes arrays:
+            let j = 32 - i - 1;
+            bytes[i] = rand1_bytes[j];
+            bytes[16 + i] = rand2_bytes[j];
+        }
+
+        all_bytes[k] = bytes;
     }
 
-    bytes
+    all_bytes
 }
 
 // We'll keep this function around, because some users might need it if their
 // shared_secret is derived in a way which is not safe for usage with the
 // poseidon2 approach.
-fn extract_close_to_uniformly_random_256_bits_from_ecdh_shared_secret_using_sha256(
+fn extract_close_to_uniformly_random_256_bits_from_ecdh_shared_secret_using_sha256<let N: u32>(
     shared_secret: Point,
-) -> [u8; 32] {
+) -> [[u8; 32]; N] {
+    assert(N == 1); // Figuring out domain separation for more than 1 would be fiddly. Doable, but fiddly.
```

Comment:
This makes little sense - this function is not generic over `N`, there's only one possible value.

Reply:
I'll refactor